src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h

```diff
@@ -205,20 +205,21 @@ class LeaseStatusTracker {
 
   /// Indicate the tracker that prepare requests are sent to a specific node.
   ///
-  /// \param node_id Id of a node where prepare request is sent.
-  /// \param bundle Bundle specification the node is supposed to prepare.
-  /// \return False if the prepare phase was already started. True otherwise.
-  bool MarkPreparePhaseStarted(const NodeID &node_id,
-                               const std::shared_ptr<const BundleSpecification> &bundle);
+  /// \param node_id Id of a node to which prepare request will be sent.
+  /// \param bundles Bundles that will be supposed to schedule.
+  void MarkPreparePhaseStarted(
+      const NodeID &node_id,
+      const std::vector<std::shared_ptr<const BundleSpecification>> &bundles);
```

Comment:
Do we need this refactoring? Is it possible to keep this as it is and only change this part? https://github.com/ray-project/ray/blob/aca954e8dd858bdd76b70af61a59a18df08b87e0/src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc#L173

I think we just need to re-organize bundles per node and send PrepareRequests PRCs with bundles?

Reply:
hmmm...  We can keep it as it is,  but the code would be like this: 
```
const auto &pending_bundles = GenUnplacedBundlesMap(bundles, selected_nodes);
for (const auto &node_to_bundles : pending_bundles) {
  const auto &node_id = node_to_bundles.first;
  const auto &bundles = node_to_bundles.second;
  for (const auto &bundle: bundles) {
    lease_status_tracker->MarkPreparePhaseStarted(node_id, bundle);
  }
    
PrepareResources(bundles, gcs_node_manager_.GetAliveNode(node_id),
                 [this, bundles, node_id, lease_status_tracker, failure_callback,
                  success_callback](const Status &status) {
                   lease_status_tracker->MarkPrepareRequestReturned(node_id, bundles,
                                                                    status);
                   if (lease_status_tracker->AllPrepareRequestsReturned()) {
                     OnAllBundlePrepareRequestReturned(
                         lease_status_tracker, failure_callback, success_callback);
                   }
                     });
```
there will be two levels for each, I just don't like this, LOL.
Is there any special reason that we should do this?

Reply:
What about we also don't refactor `MarkPrepareRequestReturned` and just change the callback semantics? We can make it return (const Status &status and const &std::shared_ptr<BundleSpecification> &bundles)

```
PrepareResources(bundles, gcs_node_manager_.GetAliveNode(node_id),
                 [this, bundles, node_id, lease_status_tracker, failure_callback,
                  success_callback](const Status &status, const &std::shared_ptr<BundleSpecification> &bundles) {
                   for (bundle : bundles) {
                       lease_status_tracker->MarkPrepareRequestReturned(node_id, bundles,
                                                                    status);
                   }
                   if (lease_status_tracker->AllPrepareRequestsReturned()) {
                     OnAllBundlePrepareRequestReturned(
                         lease_status_tracker, failure_callback, success_callback);
                   }
                     });
```

The reason why I prefer this approach is imo it seems to be cleaner to change the batch logic than `MarkPrepareRequestReturned` logic

Reply:
It seems two different methods.... `MarkPreparePhaseStarted ` and `MarkPrepareRequestReturned `

Reply:
> It seems two different methods.... MarkPreparePhaseStarted and MarkPrepareRequestReturned

What does this mean?

Reply:
just forget it, I've reverted the modification about the `lease_status_tracker `.